Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/EndRide.hs

```diff
@@ -223,30 +235,58 @@ endRide handle@ServiceHandle {..} rideId req = withLogTag ("rideId-" <> rideId.g
-      logTagInfo "driver -> endRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId rideOld.id)
-      pure driverReq.point
-    DashboardReq dashboardReq -> do
-      logTagInfo "dashboard -> endRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId rideOld.id)
-      case dashboardReq.point of
-        Just point -> pure point
-        Nothing -> do
-          pure $ getCoordinates booking.toLocation
-    CronJobReq cronJobReq -> do
-      logTagInfo "cron job -> endRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId rideOld.id)
-      case cronJobReq.point of
-        Just point -> pure point
-        Nothing -> do
-          pure $ getCoordinates booking.toLocation
-    CallBasedReq _ -> do
-      pure $ getCoordinates booking.toLocation
+  (tripEndPoint, odometerEndReading) <-
+    case rideOld.rideDetails of
+      DRide.RideDetailsOnDemand {} -> do
+        case req of
+          DriverReq driverReq -> do
+            logTagInfo "driver -> endRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId rideOld.id)
+            pure (driverReq.point, driverReq.odometerEndReading :: Maybe Centesimal) -- Irrelevant
+          DashboardReq dashboardReq -> do
+            logTagInfo "dashboard -> endRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId rideOld.id)
+            case dashboardReq.point of
+              Just point -> pure (point, Nothing)
+              Nothing -> do
+                pure (getCoordinates booking.bookingDetails.toLocation, Nothing)
+          CronJobReq cronJobReq -> do
+            logTagInfo "cron job -> endRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId rideOld.id)
+            case cronJobReq.point of
+              Just point -> pure (point, Nothing)
+              Nothing -> do
+                pure (getCoordinates booking.bookingDetails.toLocation, Nothing)
+          CallBasedReq _ -> do
+            pure (getCoordinates booking.bookingDetails.toLocation, Nothing)
+      DRide.RideDetailsRental {} -> do
+        case req of
+          DriverReq driverReq -> do
+            logTagInfo "driver -> endRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId rideOld.id)
+            case driverReq.odometerEndImage of
+              Nothing -> pure ()
+              Just image -> do
+                let merchantId = driverReq.requestor.id
+                imagePath <- createPath driverId.getId merchantId.getId rideId.getId
+                _ <- fork "S3 Put Image" $ S3.put (T.unpack imagePath) image
+                void $ QRide.updateOdometerStartReadingImagePath rideId (Just imagePath)
+            pure (driverReq.point, driverReq.odometerEndReading :: Maybe Centesimal)
+          DashboardReq dashboardReq -> do
+            logTagInfo "dashboard -> endRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId rideOld.id)
+            case dashboardReq.point of
+              Just point -> pure (point, Nothing)
```

Comment:
should be `(point, dashboardReq.odometerEndReading)` the same as for `driverReq`,  I guess

Reply:
@roman-bodavskiy  we don't have odometerEndReading in dashboard cc: @0utkarsh